cmd/frontend/webhooks/middleware.go

```diff
@@ -0,0 +1,155 @@
+	"github.com/sourcegraph/sourcegraph/internal/types"
+	"github.com/sourcegraph/sourcegraph/schema"
+)
+
+// SetExternalServiceID attaches a specific external service ID to the current
+// webhook request for logging purposes.
+func SetExternalServiceID(ctx context.Context, id int64) {
+	if setter, ok := ctx.Value(setterContextKey).(contextFunc); ok {
+		setter(id)
+	} else {
+		log15.Error("cannot get setter from context; this likely means that SetExternalServiceID has been called from outside a HTTP handler wrapped in LogMiddleware")
+	}
+}
+
+// LogMiddleware tracks webhook request content and stores it for diagnostic
+// purposes.
+type LogMiddleware struct {
+	store database.WebhookLogStore
+}
+
+// NewLogMiddleware instantiates a new LogMiddleware.
+func NewLogMiddleware(store database.WebhookLogStore) *LogMiddleware {
+	return &LogMiddleware{store}
+}
+
+func (mw *LogMiddleware) Logger(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// If logging is disabled, we'll immediately forward to the next
+		// handler, turning this middleware into a no-op.
+		if !loggingEnabled(conf.Get()) {
+			next.ServeHTTP(w, r)
+			return
+		}
+
+		// Split the body reader so we can also access it. We need to shim an
+		// io.ReadCloser implementation around the TeeReader, since TeeReader
+		// doesn't implement io.Closer, but Request.Body is required to be an
+		// io.ReadCloser.
+		type readCloser struct {
+			io.Reader
+			io.Closer
+		}
+		buf := &bytes.Buffer{}
+		tee := io.TeeReader(r.Body, buf)
+		r.Body = readCloser{tee, r.Body}
+
+		// Set up a logging response writer so we can also store the response;
+		// most importantly, the status code.
+		writer := &responseWriter{
+			ResponseWriter: w,
+		}
+
+		// The external service ID is looked up within the webhook handler, but
+		// we need access to it to be able to store the webhook log with the
+		// appropriate external service ID. To handle this, we'll put a setter
+		// closure into the context that can then be used by
+		// SetExternalServiceID to receive the external service ID from the
+		// handler.
+		var externalServiceID *int64
+		var setter contextFunc = func(id int64) {
+			externalServiceID = &id
+		}
+		ctx := context.WithValue(r.Context(), setterContextKey, setter)
+
+		// Delegate to the next handler.
+		next.ServeHTTP(writer, r.WithContext(ctx))
+
+		// Write the payload.
+		if err := mw.store.Create(r.Context(), &types.WebhookLog{
+			ExternalServiceID: externalServiceID,
+			StatusCode:        writer.statusCode,
+			Request: types.WebhookLogMessage{
+				Header: r.Header,
+				Body:   buf.Bytes(),
+			},
+			Response: types.WebhookLogMessage{
+				Header: writer.Header(),
+				Body:   writer.buf.Bytes(),
+			},
+		}); err != nil {
+			// This is non-fatal, but almost certainly indicates a significant
+			// problem nonetheless.
+			log15.Warn("error writing webhook log", "err", err)
```

Comment:
Significant error sounds like this warrants: 😁 
```suggestion
			log15.Error("error writing webhook log", "err", err)
```

Reply:
I went back and forth on the level of this: it's significant, but it's also non-critical in the sense that Sourcegraph proper will operate fine if it occurs; you just lose the webhook logging capability.

I don't feel strongly about this, though, so let's round up to `Error`.

Reply:
I don't have a strong preference either. Your call :)